### DISCUSSION TITLE
Streaming results from Python?

### DISCUSSION THREAD

**Discussion by wisp3rwind** [duckdb/duckdb#4941] [Q&A]
I've just realized that the Python API seems to always fully materialize query results, which easily leads to OOM issues.

What I need to do is essentially
```python
for batch in rel.record_batch(batch_size):
    # process batch further using numpy, in a way that is not easily pushed down into duckdb
```
but, counter-intuitively, `record_batch` seems to first build the whole result set in memory (up to several GB in this case).

Is it in fact correct that there's no way to iterate over results from Python without fully materializing them?

**Comment by Alex-Monahan** [comment]
Have you experimented with fetchmany?
https://github.com/duckdb/duckdb/pull/4465

You can also use fetch_df_chunk!
https://github.com/duckdb/duckdb/pull/1366

**Comment by wisp3rwind** [comment]
Thanks for the quick reply!

---

No, I haven't tried those, but I don't see how there can be a difference: Stripping down the query to
```python
rel = con.from_query(f"""SELECT
    *
FROM parquet_scan(
    '{dset_dir}/*/*/energies-*.parquet',
    HIVE_PARTITIONING=1
)
""")

res = rel.execute()
```

the complete result set appears to be allocated already (as evidenced by RSS increasing by a few GB)  in the `execute()` call, so using `fetchmany`, `fetch_df_chunk` or `fetch_arrow_reader` afterwards won't make a difference.

For this simple query (and also the one I'm actually running, which includes a few filters and computations), there should be no need to build the full result before starting to yield result chunks. The only complication here might be the use of `HIVE_PARTITIONING`, I'll have to check whether querying individual files makes a difference.

---

**EDIT:** Same behaviour if just using `parquet_scan('energies-001.parquet')`, i.e. without `HIVE_PARTITIONING`.

**Comment by Alex-Monahan** [comment]
Hmm, I'm not sure I have a great answer then. One workaround would be to insert your results into a table, then read from that using LIMIT and OFFSET. However, that is not very optimized I believe.

**Reply by wisp3rwind** [reply]
In fact, I do have a workaround that works well enough: I can loop over certain groups of the data, and `filter()` the query accordingly in each iteration. That's reasonably efficient, since the group key is constant in each parquet row group. Thus, the filter can be resolved using only the parquet metadata. It does increase the complexity of the code, which I was hoping to avoid.

For context, this is data analysis I'm running on a laptop (which in principle should take less than 1 minute), and I might not have sufficient RAM to fit the full result set. Thus, for some bigger data sets, I've been encountering OOM issues recently. It took me quite a while to pin them down to duckdb, since I expected (maybe wrongly) that the data should be streamed. Moving to a bigger machine would in principle be possible, but is not really necessary for this workload and would be quite inconvenient.

Can you tell what the intended behaviour is here? The existence of `fetchmany`, `fetch_df_chunk` and `fetch_arrow_reader` suggests that the result should in fact be allocated in chunks. Is this simply an issue of something not being implemented (yet), or is there a bug somewhere that forces full result materialization?

I've browsed the source code a bit, and if I'm understanding things correctly, this boils down to whether the `allow_stream_result` flag ends up being set for a query. However, I couldn't tell quickly whether using a table function such as `parquet_scan` from Python would have any effect on this flag.

**Reply by Mause** [reply]
> In fact, I do have a workaround that works well enough: I can loop over certain groups of the data, and `filter()` the query accordingly in each iteration. That's reasonably efficient, since the group key is constant in each parquet row group. Thus, the filter can be resolved using only the parquet metadata. It does increase the complexity of the code, which I was hoping to avoid.
> 
> For context, this is data analysis I'm running on a laptop (which in principle should take less than 1 minute), and I might not have sufficient RAM to fit the full result set. Thus, for some bigger data sets, I've been encountering OOM issues recently. It took me quite a while to pin them down to duckdb, since I expected (maybe wrongly) that the data should be streamed. Moving to a bigger machine would in principle be possible, but is not really necessary for this workload and would be quite inconvenient.
> 
> Can you tell what the intended behaviour is here? The existence of `fetchmany`, `fetch_df_chunk` and `fetch_arrow_reader` suggests that the result should in fact be allocated in chunks. Is this simply an issue of something not being implemented (yet), or is there a bug somewhere that forces full result materialization?
> 
> I've browsed the source code a bit, and if I'm understanding things correctly, this boils down to whether the `allow_stream_result` flag ends up being set for a query. However, I couldn't tell quickly whether using a table function such as `parquet_scan` from Python would have any effect on this flag.

Is this something you'd be able to provide a reproduction for? We should only be materializing 1024 rows at once (the current vector size)

**Reply by wisp3rwind** [reply]
Can be as simple as (duckdb 0.5.1, Python 3.9, Linux):

Scripty `gen.py` to build a very simple, but somewhat big table
```python
#! /usr/bin/env python

from numpy.random import Generator, PCG64
import pyarrow as pa
import pyarrow.parquet as pq

# Init rng in a reproducible way
rng = Generator(PCG64(12345))

# row count, big enough to make larger than intended memory usage obvious
N = 100_000_000

columns = {}
columns["random_f64"] = pa.array(rng.random(N))

table = pa.Table.from_pydict(columns)

pq.write_table(
    table,
    "test.parquet",
    compression="zstd",
    # This is also pyarrows default: A data page corresponds to the smallest chunk that
    # duckdb can read.
    data_page_size=1024 * 1024,  # bytes
    use_dictionary=False,
)


```

Run and inspect:

```
$ ./gen.py
$ du -h test.parquet
717M	test.parquet
$ parquet-tools inspect --detail test.parquet | grep -E -A1 '(RowGroup|PageEncodingStats)'
        RowGroup
            columns = list
--
                            PageEncodingStats
                                count = 512
--
        RowGroup
            columns = list
--
                            PageEncodingStats
                                count = 251
```
i.e. the file indeed consists of many data pages (512 + 152 in 2 row groups, thus ~ 1MB per page as requested).

Read the whole file back using `read.py`:
```python
#! /usr/bin/env python

import duckdb
import gc
from resource import getrusage, RUSAGE_SELF
# Add some sleep calls to have a chance to observe this in (h)top
import time


# getrusage(RUSAGE_SELF).ru_maxrss gives the maximum RSS of the current process
# units depend on the OS, see `man getrusage`. kB on Linux, but according to
# stackoverflow, bytes on Mac


con = duckdb.connect()
# baseline
print("Max. RSS after connect: {}".format(getrusage(RUSAGE_SELF).ru_maxrss))


rel = con.query("""
    SELECT *
    FROM parquet_scan('test.parquet');
""")
# just building the query, memory usage stays (almost) constant 
print("Max. RSS after query relation: {}".format(getrusage(RUSAGE_SELF).ru_maxrss))
time.sleep(10)

res = rel.execute()
# max RSS increases by 800MB, whereas I'd expect little change until actually
# fetching any data.
print("Max. RSS after query exec: {}".format(getrusage(RUSAGE_SELF).ru_maxrss))
time.sleep(10)

while True:
    data = res.fetch_df_chunk()
    if data.empty:
        break
# Max RSS increases to 3.4GB (probably pandas overhead, and GC not kicking in
# early enough to collect previous chunks?)
print("Max. RSS after result fetch: {}".format(getrusage(RUSAGE_SELF).ru_maxrss))
time.sleep(10)
```

```
$ ./read.py
Max. RSS after connect: 50560
Max. RSS after query relation: 50560
Max. RSS after query exec: 867308
Max. RSS after result fetch: 3447948
```

**Reply by Mause** [reply]
Yeah I see what you mean, that isn't great at all. @Tishj probably worth having a look at this?

**Reply by Mytherin** [reply]
The reason streaming results are disabled in the Python client is that streaming results disable parallel execution of the final pipeline. When streaming results are enabled the final pipeline is moved to the client, and chunks are lazily materialized when `Fetch` is called in a single-threaded manner.

The problem with that is that the Python client is currently set-up in such a manner that we do not know what type of result fetching the user wants to use when constructing the result. The query is executed when `execute` is called. That means that simply enabling streaming results will disable parallelism when actual full materialization is desired, which is of course not ideal, especially since many times `fetchdf` or `fetchall` will be used instead.

In order to make streaming work well we need to postpone executing the query until a fetch method is called, and use either a streaming or a materialized query result depending on the fetch method that is used:

* `fetch_one, fetchmany, fetch_df_chunk` -> streaming
* `fetchall, fetchnumpy, fetchdf, ... (all others)` -> materialized

**Reply by Mytherin** [reply]
Actually thinking about this some more - I propose we remove the `DuckDBPyResult` class altogether, and have `execute` output a `DuckDBPyRelation` node. That would avoid requiring the need to duplicate methods, and would make our interface significantly more consistent.

**Reply by wisp3rwind** [reply]
Hi @Mytherin, thanks for the detailed explanation! That refactoring of the `DuckDBPyResult` sounds sensible from my (limited) point of view. Looking forward to seing that implemented (eventually)!

**Reply by wisp3rwind** [reply]
I've just seen https://github.com/duckdb/duckdb/pull/5597, nice!

Since I couldn't tell whether this would already allow for streaming `RecordBatchReader`s, I tried to run my example code from above again. That didn't work, since the `fetch_df_chunk` method doesn't exist anymore. Was that removed intentionally?

With the slightly changed `read.py`

```python
#! /usr/bin/env python

import duckdb
import gc
from resource import getrusage, RUSAGE_SELF
# Add some sleep calls to have a chance to observe this in (h)top
import time


# getrusage(RUSAGE_SELF).ru_maxrss gives the maximum RSS of the current process
# units depend on the OS, see `man getrusage`. kB on Linux, but according to
# stackoverflow, bytes on Mac


con = duckdb.connect()
# baseline
print("Max. RSS after connect: {}".format(getrusage(RUSAGE_SELF).ru_maxrss))


rel = con.query("""
    SELECT *
    FROM parquet_scan('test.parquet');
""")
# just building the query, memory usage stays (almost) constant 
print("Max. RSS after query relation: {}".format(getrusage(RUSAGE_SELF).ru_maxrss))
time.sleep(10)

reader = rel.fetch_arrow_reader(batch_size=16_384)
print(type(reader))
# max RSS increases by 800MB, whereas I'd expect little change until actually
# fetching any data.
print("Max. RSS after query exec: {}".format(getrusage(RUSAGE_SELF).ru_maxrss))
time.sleep(10)

for batch in reader:
    batch.to_pydict()
# Max RSS increases only slightly, as expected.
print("Max. RSS after result fetch: {}".format(getrusage(RUSAGE_SELF).ru_maxrss))
time.sleep(10)
```

```
$ ./read.py
Max. RSS after connect: 32624
Max. RSS after query relation: 33068
<class 'pyarrow.lib.RecordBatchReader'>
Max. RSS after query exec: 878760
Max. RSS after result fetch: 880900
```

however, it seems that streaming is still not possible (note that after the `fetch_arrow_reader(batch_size=16_384)` call, the entire result set appears to be in memory).

**Reply by wisp3rwind** [reply]
Forgot to mention, this is duckdb 0.7.0 on Python 3.11.1

**Reply by Mytherin** [reply]
Indeed the streaming result set is not yet implemented - but indeed with the merging of #5597 this is significantly easier to do. I will push a PR. 

As an aside - you should import `pyarrow` before to measure this correctly:

```py
#! /usr/bin/env python

import duckdb
import gc
import pyarrow
from resource import getrusage, RUSAGE_SELF
# Add some sleep calls to have a chance to observe this in (h)top
import time


# getrusage(RUSAGE_SELF).ru_maxrss gives the maximum RSS of the current process
# units depend on the OS, see `man getrusage`. kB on Linux, but according to
# stackoverflow, bytes on Mac


con = duckdb.connect()
# baseline
print("Max. RSS after connect: {}".format(getrusage(RUSAGE_SELF).ru_maxrss))


rel = con.query("""
    SELECT *
    FROM parquet_scan('test.parquet');
""")
# just building the query, memory usage stays (almost) constant 
print("Max. RSS after query relation: {}".format(getrusage(RUSAGE_SELF).ru_maxrss))

reader = rel.fetch_arrow_reader(batch_size=16_384)
print(type(reader))
# max RSS increases by 800MB, whereas I'd expect little change until actually
# fetching any data.
print("Max. RSS after query exec: {}".format(getrusage(RUSAGE_SELF).ru_maxrss))

for batch in reader:
    batch.to_pydict()
# Max RSS increases only slightly, as expected.
print("Max. RSS after result fetch: {}".format(getrusage(RUSAGE_SELF).ru_maxrss))

```

Otherwise `fetch_arrow_reader` imports `pyarrow`, counting towards the increase in memory usage there.

**Reply by wisp3rwind** [reply]
> Indeed the streaming result set is not yet implemented - but indeed with the merging of #5597 this is significantly easier to do. I will push a PR.

Thanks a lot!

> As an aside - you should import `pyarrow` before to measure this correctly:
> 
> [...]
> 
> Otherwise `fetch_arrow_reader` imports `pyarrow`, counting towards the increase in memory usage there.

Agreed, that accounts for about 30MB. For this particular test (and due the much larger size of the result set in comparison), it doesn't really matter, though.

**Reply by oridag** [reply]
@Mytherin was streaming support ever implemented?

**Reply by oridag** [reply]
Looks like it was in this [PR](https://github.com/duckdb/duckdb/pull/6346). But doing an Iceberg scan, I still see buffering.
Is there a way to do a streaming query from Iceberg?
